// @shd101wyy/mume/out/src/markdown-engine-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultMarkdownEngineConfig = void 0;
exports.defaultMarkdownEngineConfig = {
    configPath: null,
    usePandocParser: false,
    breakOnSingleNewLine: true,
    enableTypographer: false,
    enableWikiLinkSyntax: true,
    enableLinkify: true,
    enableEmojiSyntax: true,
    enableExtendedTableSyntax: false,
    enableCriticMarkupSyntax: false,
    useGitHubStylePipedLink: true,
    wikiLinkFileExtension: ".md",
    protocolsWhiteList: "http://, https://, atom://, file://, mailto:, tel:",
    mathRenderingOption: "KaTeX",
    mathInlineDelimiters: [
        ["$", "$"],
        ["\\(", "\\)"],
    ],
    mathBlockDelimiters: [
        ["$$", "$$"],
        ["\\[", "\\]"],
    ],
    mathRenderingOnlineService: "https://latex.codecogs.com/gif.latex",
    codeBlockTheme: "auto.css",
    previewTheme: "github-light.css",
    revealjsTheme: "white.css",
    mermaidTheme: "default",
    frontMatterRenderingOption: "none",
    imageFolderPath: "/assets",
    printBackground: false,
    chromePath: "",
    imageMagickPath: "",
    pandocPath: "pandoc",
    pandocMarkdownFlavor: "markdown-raw_tex+tex_math_single_backslash",
    pandocArguments: [],
    latexEngine: "pdflatex",
    enableScriptExecution: false,
    enableHTML5Embed: false,
    HTML5EmbedUseImageSyntax: true,
    HTML5EmbedUseLinkSyntax: false,
    HTML5EmbedIsAllowedHttp: false,
    HTML5EmbedAudioAttributes: 'controls preload="metadata"',
    HTML5EmbedVideoAttributes: 'controls preload="metadata"',
    puppeteerWaitForTimeout: 0,
    usePuppeteerCore: true,
    puppeteerArgs: [],
    plantumlServer: "",
};
