
<!DOCTYPE html>
<!--
    Plain-Academic by Vasilios Mavroudis
    Released under the Simplified BSD License/FreeBSD (2-clause) License.
    https://github.com/mavroudisv/plain-academic
-->

<html lang="en">
<head>
  <title>Samuel Sledzieski</title>
  <link rel="icon" 
  type="image/png" 
  href="img/favicon.ico">
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <link rel="stylesheet" href="style.css">
  <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.0/jquery.min.js"></script>
  <link href='https://fonts.googleapis.com/css?family=Oswald:700' rel='stylesheet' type='text/css'>
  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css" integrity="sha384-Gn5384xqQ1aoWXA+058RXPxPg6fy4IWvTNh0E263XmFcJlSAwiGgFAW/dAiS6JXm" crossorigin="anonymous">
  
  <meta charset="utf-8" />
</head>
<body>


<!-- Navigation -->
<nav class="navbar navbar-expand-lg navbar-light bg-light">
    <a class="navbar-brand" href="./index.html">Samuel Sledzieski</a>

    <div class="collapse navbar-collapse" id="navbarSupportedContent">
        <ul class="navbar-nav mr-auto">
        <li class = "nav-item">
            <a class="nav-link" href="./index.html">Home</span></a>
        </li>
        <li class = "nav-item active">
            <a class="nav-link" href="./publications.html">Publications</a>
        </li>
        <li class = "nav-item"></li>
            <a class="nav-link" href="./software.html">Software</a>
        </li>
        <!-- <li class = "nav-item">
            <a class="nav-link" href="./notebook.html">Lab Notebook</a>
        </li> -->
        <li class="nav-item">
            <a class="nav-link" href="files/Samuel_Sledzieski_CV.pdf" target="_blank">CV</a>
        </li>
        </ul>
    </div>
</nav>
  
<!-- Body -->
<br>
<div class="container-fluid">

    <div class="row">

        <!-- Contact Info on the Sidebar -->
        <div class="col-md-3">
            <!-- Main Image -->
            <img class="img-responsive" width = "200" src="img/sledzieski_samuel_profile.jpg" alt=""><br>
            
            <div style="font-family: 'Oswald', sans-serif; font-size: 32px;"><b>Samuel Sledzieski</b></div><br>
            <p>
                Computational Biology PhD Student<br>
                Massachusetts Institute of Technology, USA<br> 
            </p>
            <p>
                Email: <b><a href="mailto:samsl@mit.edu">samsl@mit.edu</a></b><br>
                GitHub: <b><a href="https://github.com/samsledje/">samsledje</a></b><br>
                Twitter: <b><a href="https://twitter.com/samsledzieski/">@samsledzieski</a></b><br>
                <b><a href="https://scholar.google.com/citations?user=BoZ7AqYAAAAJ&hl=en">Google Scholar</a></b><br>
            </p>
            <!-- <p> <a href="https://scholar.google.com/citations?user=BoZ7AqYAAAAJ&hl=en"><strong>Google Scholar</strong></a> | 
                <a href="https://www.linkedin.com/in/samuelsledzieski/"><strong>Linkedin</strong></a> | 
                <a href="https://github.com/samsledje/"><strong>GitHub</strong></a> |
                <a href="https://twitter.com/samsledzieski/"><strong>Twitter</strong></a></p> -->
        </div>

        <!-- Page Content -->
        <div class="col-md-8" style="height:auto">    
            <h2 id="selected">Publications</h2>

            <p><b>*Equal Contribution</b></p>

            <h4>Journal</h4>
            <ul>
                <li><b>Sledzieski*</b>, Devkota*, Singh, Cowen, Berger, "TT3D: Leveraging Pre-Computed Protein Sequence Models to Predict Protein-Protein Interactions,"
                    Bioinformatics, 2023: btad663.
                    <a href="https://academic.oup.com/bioinformatics/article/39/11/btad663/7332153" target="_blank"><b>[PDF]</b></a>
                    <a href="https://cb.csail.mit.edu/cb/topsyturvy/" target="_blank"><b>[Software]</b></a>
                </li>
                <li><b>Sledzieski*</b>, Singh*, Bryson, Cowen, Berger, "Contrastive learning in protein language space predicts interactions between drugs and protein targets",
                    Proceedings of the National Academy of Sciences 120.24 (2023): e2220778120.
                    <a href="https://www.pnas.org/doi/epdf/10.1073/pnas.2220778120" target="_blank"><b>[PDF]</b></a>
                    <a href="https://cb.csail.mit.edu/cb/conplex/" target="_blank"><b>[Software]</b></a>
                </li>
                <li>Kumar, Brenner, <b>Sledzieski</b>, Olaosebikan, Lynn-Goin, Putnam, Yang, Lewinski, Singh, Daniels,
                    Cowen, Klein-Seetharaman, "Transfer of knowledge from model organisms to evolutionarily distant non-model organisms:
                    The coral <i>Pocillopora damicornis</i> membrane signaling receptome," Plos one 18.2 (2023). 10.1371/journal.pone.0270965
                    <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9897584/pdf/pone.0270965.pdf" target="_blank"><b>[PDF]</b></a>
                </li>
                <li>Zaman*, <b>Sledzieski*</b>, Wu, Bansal, "virDTL: Viral recombination analysis through phylogenetic reconciliation and its 
                    application to sarbecoviruses and SARS-CoV-2," J Comput Biol. 2022 Sep 20. doi: 10.1089/cmb.2021.0507. Epub ahead of print. PMID: 36125448.
                    <a href="https://www.liebertpub.com/doi/10.1089/cmb.2021.0507" target="_blank"><b>[PDF]</b></a>
                    <a href="https://github.com/suz11001/virDTL" target="_blank"><b>[Software]</b></a>
                </li>
                <li>Singh*, Devkota*, <b>Sledzieski</b>, Berger, Cowen, "Topsy-Turvy: integrating a global view into sequence-based PPI prediction,"
                    Bioinformatics, 38.Supplement_1 (July 2022): i264–i272.
                    <a href="https://academic.oup.com/bioinformatics/article/38/Supplement_1/i264/6617505" target="_blank"><b>[PDF]</b></a>
                    <a href="https://cb.csail.mit.edu/cb/topsyturvy/" target="_blank"><b>[Software]</b></a>
                </li>
                <li><b>Sledzieski*</b>, Singh*, Cowen, Berger, “D-SCRIPT translates genome to phenome with
                    sequence-based, structure-aware, genome-scale predictions of protein-protein interactions,”
                    Cell Systems 12.10 (2021): 969-982.
                    <a href="https://www.cell.com/cell-systems/pdfExtended/S2405-4712(21)00333-1" target="_blank"><b>[PDF]</b></a>
                    <a href="https://cb.csail.mit.edu/cb/dscript/" target="_blank"><b>[Software]</b></a>
                </li>
            </ul>

            <h4>Conference</h4>
            <ul>
                <li><b>Sledzieski</b>, Kshirsagar, Baek, Berger, Dodhia, Lavista Ferres, "Parameter-Efficient Fine-Tuning of
                    Protein Language Models Improves Prediction of Protein-Protein Interactions",
                    Machine Learning for Structural Biology Workshop at NeurIPS, December 2023.
                    <!-- <a href="https://www.biorxiv.org/content/10.1101/2022.11.03.515086v1.full.pdf" target="_blank"><b>[PDF]</b></a> -->
                </li>
                <li><b>Sledzieski*</b>, Singh*, Cowen, Berger, "Contrasting drugs from decoys",
                    Machine Learning for Structural Biology Workshop at NeurIPS, December 2022.
                    <a href="https://www.biorxiv.org/content/10.1101/2022.11.03.515086v1.full.pdf" target="_blank"><b>[PDF]</b></a>
                </li>
                <li><b>Sledzieski*</b>, Singh*, Cowen, Berger, "Adapting protein language models for rapid DTI prediction",
                    Machine Learning for Structural Biology Workshop at NeurIPS, December 2021.
                    <a href="https://www.biorxiv.org/content/10.1101/2022.11.03.515084v1.full.pdf" target="_blank"><b>[PDF]</b></a>
                </li>
                <li><b>Sledzieski*</b>, Singh*, Cowen, Berger, “Sequence-based prediction of protein-protein interactions:
                    a structure-aware interpretable deep learning model,” Conference on Research in Computational
                    Molecular Biology (RECOMB), August 2021.
                    <a href="https://www.biorxiv.org/content/10.1101/2021.01.22.427866v1.full.pdf" target="_blank"><b>[PDF]</b></a>
                </li>
                <li><b>Sledzieski</b>, Zhang, Mandoiu, Bansal, “TreeFix-TP: Phylogenetic Error Correction for Accurate
                    Reconstruction of Viral Transmission Networks,” Pacific Symposium on Biocomputing (PSB), January 2021.
                    Proceedings, pages 119-130.
                    <a href="https://www.worldscientific.com/doi/pdf/10.1142/9789811232701_0012" target="_blank"><b>[PDF]</b></a>
                    <a href="https://github.com/samsledje/TreeFix-TP" target="_blank"><b>[Software]</b></a>
                </li>
            </ul>

            <h4>Preprints</h4>
            <ul>
                <li><b>Sledzieski</b>, Kshirsagar, Baek, Berger, Dodhia, Lavista Ferres, “Democratizing Protein Language Models with 
                    Parameter-Efficient Fine-Tuning,” <i>Under review.</i> 
                    <i>Conference on Research in Computational Molecular Biology</i> (RECOMB)
                    <a href="https://www.biorxiv.org/content/10.1101/2023.11.09.566187v1.full.pdf" target="_blank"><b>[PDF]</b></a>
                </li>
                <li>Kousi, Boix, Park, Mathys, <b>Sledzieski</b>, Peng, Bennett, Tsai, Kellis, “Single-cell mosaicism analysis
                    reveals cell-type-specific somatic mutational burden in Alzheimer’s Dementia,” bioRxiv. posted 22
                    April 2022, 10.1101/2022.04.21.489103.
                    <a href="https://www.biorxiv.org/content/10.1101/2022.04.21.489103v1.full.pdf" target="_blank"><b>[PDF]</b></a>
                </li>
            </ul>

        <!-- Other people may like it too! -->
        <a style="color:#b5bec9;font-size:0.8em; float:left;" href="https://github.com/mavroudisv/plain-academic">Website design courtesy of Vasilios Mavroudis</a>
        </div>

    </div>

</div>

    
</body>

</html>
