(self.webpackChunk_N_E=self.webpackChunk_N_E||[]).push([[327],{7468:function(e,s,a){(window.__NEXT_P=window.__NEXT_P||[]).push(["/projects",function(){return a(8319)}])},8319:function(e,s,a){"use strict";a.r(s),a.d(s,{__toc:function(){return h}});var t=a(5893),n=a(2673),i=a(373),r=a(14);a(7954);var o=a(2643);let h=[];function c(e){let s=Object.assign({p:"p",a:"a"},(0,o.a)(),e.components);return(0,t.jsxs)("div",{children:[(0,t.jsxs)(s.p,{children:[(0,t.jsx)("h2",{children:"Genomic Basis of Adaptive Evolution"}),"\nStickleback repeated adaptation to new environments provide a powerful mechanism to study the molecular genetics of evolutionary changes in wild populations. As we develop stickleback genomic sequencing technologies, we can start mapping genes that control evolutionary change to the genetic basis of adaptive evolution."]}),(0,t.jsxs)(s.p,{children:[(0,t.jsx)("h2",{children:"Marine Stickleback Genome"}),"\nThe newly minted marine stickleback genome was assembled into 22 chromosomes, 455MB in genome length with an N50 of 21,246,911 base pairs and nearly gapless with only 715 gaps remaining. The marine stickleback was from the Rabbit Slough, Alaska population was was sequenced by combining several long read sequencing technologies with comprehensive enhanced informative short read Illumina sequencing."]}),(0,t.jsxs)("li",{children:["Visualization: ",(0,t.jsx)(s.a,{href:"http://genome.ucsc.edu/cgi-bin/hgHubConnect?hgHub_do_redirect=on&hgHubConnect.remakeTrackHub=on&hgHub_do_firstDb=1&hubUrl=http://trackhub.genome.duke.edu/lowelab/edotau/rabsTHREEspine/myHub/hub.txt",children:"UCSC Genome Browser"})]}),(0,t.jsxs)("li",{children:["Chain Nets to other stickleback assemblies are available: ",(0,t.jsx)(s.a,{href:"https://edotau.github.io/chainNets",children:"Here"})]}),(0,t.jsxs)("li",{children:["Stickleback Assembly: ",(0,t.jsx)(s.a,{href:"http://trackhub.genome.duke.edu/lowelab/edotau/rabsTHREEspine/rabsTHREEspine.fa",children:"Download"})]}),(0,t.jsxs)("li",{children:["Raw Fastq Data Files: ",(0,t.jsx)(s.a,{href:"http://trackhub.genome.duke.edu/lowelab/edotau/rabsTHREEspine/raw_data/assembly/pacbio/",children:"PacBio"})," , ",(0,t.jsx)(s.a,{href:"http://trackhub.genome.duke.edu/lowelab/edotau/rabsTHREEspine/raw_data/assembly/10xGenomics",children:"10x Genomics"})," , ",(0,t.jsx)(s.a,{href:"http://trackhub.genome.duke.edu/lowelab/edotau/rabsTHREEspine/raw_data/assembly/hic",children:"Hi-C"})," , ",(0,t.jsx)(s.a,{href:"http://trackhub.genome.duke.edu/lowelab/edotau/rabsTHREEspine/raw_data/assembly/nanopore",children:"Nanopore"})," "]}),(0,t.jsx)("h2",{children:"Stickleback Genes and Transcripts"}),(0,t.jsxs)("li",{children:["Ensembl Mapped: ",(0,t.jsx)(s.a,{href:"http://trackhub.genome.duke.edu/lowelab/edotau/rabsTHREEspine/myHub/genes/genePred/rabsTHREEspine.ensembl.mapped.gp",children:"genePred"})," , ",(0,t.jsx)(s.a,{href:"http://trackhub.genome.duke.edu/lowelab/edotau/rabsTHREEspine/myHub/genes/gtf/rabsTHREEspine.ensembl.mapped.gtf",children:"GTF"})," , ",(0,t.jsx)(s.a,{href:"http://trackhub.genome.duke.edu/lowelab/edotau/rabsTHREEspine/myHub/genes/gff/rabsTHREEspine.ensembl.mapped.gff",children:"GFF3"})," "]}),(0,t.jsxs)("li",{children:["RNASEQ Genes: ",(0,t.jsx)(s.a,{href:"http://trackhub.genome.duke.edu/lowelab/edotau/rabsTHREEspine/myHub/genes/genePred/rna-seq.genes.mapped.ensembl.gp",children:"genePred"})," , ",(0,t.jsx)(s.a,{href:"http://trackhub.genome.duke.edu/lowelab/edotau/rabsTHREEspine/myHub/genes/gtf/rna-seq.genes.mapped.ensembl.gtf",children:"GTF"})," , ",(0,t.jsx)(s.a,{href:"http://trackhub.genome.duke.edu/lowelab/edotau/rabsTHREEspine/myHub/genes/gff/rna-seq.genes.mapped.ensembl.gff",children:"GFF3"})," "]}),(0,t.jsxs)(s.p,{children:[(0,t.jsx)("h2",{children:"Genome-Wide Variation Between Freshwater and Marine Stickleback"}),"\nPairwise maine and freshwater stickleback genome alignments were produced in ",(0,t.jsx)(s.a,{href:"https://genome.ucsc.edu/goldenPath/help/axt.html",children:"AXT"})," format with ",(0,t.jsx)(s.a,{href:"http://www.bx.psu.edu/~rsharris/lastz/README.lastz-1.04.03.html",children:"LASTZ"})," and converted to ",(0,t.jsx)(s.a,{href:"https://genome.ucsc.edu/goldenPath/help/chain.html",children:"CHAIN"})," format with ",(0,t.jsx)(s.a,{href:"https://github.com/ENCODE-DCC/kentUtils/blob/master/src/hg/mouseStuff/axtChain/axtChain.c",children:"axtChain"})," . AXT and CHAIN alignments were generated in both directions. The set of alignments were computed using the marine stickleback as the target genome which used the freshwater stickleback as the query genome. However, the second set of alignments were computed in reverse, this time with freshwater genome as the target with the marine stickleback as its query genome and further processed using ",(0,t.jsx)(s.a,{href:"https://github.com/ENCODE-DCC/kentUtils/blob/master/src/hg/mouseStuff/chainSwap/chainSwap.c",children:"chainSwap"})," . LASTZ and axtChain were programmed into a wrapping ",(0,t.jsx)(s.a,{href:"https://raw.githubusercontent.com/edotau/sticklebackCipher/master/lastz/lastz.sh",children:"script"})," . All chains were merged and sorted with ",(0,t.jsx)(s.a,{href:"https://github.com/ENCODE-DCC/kentUtils/blob/master/src/hg/mouseStuff/chainMergeSort/chainMergeSort.c",children:"chainMergeSort"})," . Finally the best alignments from chains are selected with ",(0,t.jsx)(s.a,{href:"https://github.com/ENCODE-DCC/kentUtils/blob/master/src/hg/mouseStuff/chainNet/chainNet.c",children:"chainNet"})," using a red-black trees algorithm to keep track of areas of a chromosome are already covered until there are no bases left and distinguish duplicated from non-duplicated regions. The resulting file is hierarchical collection of chains, with the highest-scoring non-overlapping chains on top, and gaps filled by possible lower-scoring chains called a NET. Nets are single-coverage target genomes, but not for query genomes unless it has been filtered to be single-coverage on both target and query. We generated reciprocal-best netted chains our pairwise netted chains by writing a wrapping ",(0,t.jsx)(s.a,{href:"https://raw.githubusercontent.com/edotau/sticklebackCipher/master/lastz/reciprocalBestChainNets.sh",children:"bash script"})," . All programs to process chains and nets are open sourced UCSC Kent Utilities developed to examines genomic duplications, deletions, and rearrangements of the first whole-genome comparisons between human and mouse ",(0,t.jsx)(s.a,{href:"https://www.pnas.org/content/pnas/100/20/11484.full.pdf",children:"J Kent 2003"}),"."]}),(0,t.jsx)("h3",{children:"Marine Freshwater Hybrid Crosses"}),(0,t.jsxs)(s.p,{children:[(0,t.jsxs)("li",{children:["GATK: ",(0,t.jsx)(s.a,{href:"http://trackhub.genome.duke.edu/lowelab/edotau/rabsTHREEspine/myHub/vcfs/snps",children:"SNPS"}),", ",(0,t.jsx)(s.a,{href:"http://trackhub.genome.duke.edu/lowelab/edotau/rabsTHREEspine/myHub/vcfs/indels",children:"INDELS"})]}),"\nTwo different F1 hybrid (Freshwater x Marine) ",(0,t.jsx)(s.a,{href:"https://en.wikipedia.org/wiki/Three-spined_stickleback",children:"stickleback"})," fish were crossed from ",(0,t.jsx)(s.a,{href:"https://www.google.com/maps/place/Matadero+Creek/@37.4241044,-122.1510844,14z/data=!3m1!4b1!4m5!3m4!1s0x808fbaf41f10715b:0xc780fcc2d1157053!8m2!3d37.4241062!4d-122.1335748",children:"Matadero Creek"})," with ",(0,t.jsx)(s.a,{href:"https://www.google.com/maps/place/Little+Campbell+River/@49.0339164,-122.7053782,13z/data=!3m1!4b1!4m5!3m4!1s0x5485c5e345db5175:0x48382b94481b427a!8m2!3d49.0339216!4d-122.6703588",children:"Little Campbell"})," and ",(0,t.jsx)(s.a,{href:"https://www.google.com/maps/place/Rabbit+Slough/@61.5398359,-149.3153327,14z/data=!4m5!3m4!1s0x56c8e14ff29fb997:0x4542770c26b5618a!8m2!3d61.5333333!4d-149.2666666",children:"Rabbit Slough"})," with ",(0,t.jsx)(s.a,{href:"https://www.google.com/maps/place/Matanuska+Lake/@61.555098,-149.2414204,15z/data=!3m1!4b1!4m5!3m4!1s0x56c8e17081d30ab5:0x911087de19454c0b!8m2!3d61.5550708!4d-149.231791",children:"Lake Matanuska"}),". A Summary Table of populations sampled and libraries sequenced to investigate each allele are described in panel B. We generated ",(0,t.jsx)(s.a,{href:"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4374986/",children:"ATAC-Seq"})," , ",(0,t.jsx)(s.a,{href:"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2949280/",children:"RNA-Seq"}),", and ",(0,t.jsx)(s.a,{href:"https://www.internationalgenome.org/sites/1000genomes.org/files/docs/nature09534.pdf",children:"WGS"})," libraries to screen for allele specific enhancers near regions of increased chromatin accessibility that could perhaps be regulating differential gene expression. We perform a ",(0,t.jsx)(s.a,{href:"https://en.wikipedia.org/wiki/Fisher%27s_exact_test",children:"Fisher's Exact Test"})," genome-wide to determine statistical significance if heterogygous SNPs sites are differentially increased or decreased in chromatin accessibility are enhancers or regulatory elements nearby differentially expressed genes."]}),(0,t.jsx)("h2",{children:"Differential Chromatin Accessibility and Gene Expression"}),(0,t.jsxs)("li",{children:["Results: ",(0,t.jsx)(s.a,{href:"http://trackhub.genome.duke.edu/lowelab/edotau/rabsTHREEspine/myHub/beds/differentialChromatin/",children:"Download"})]}),(0,t.jsx)("img",{src:"/pictures/pilotF1Hybrid.svg",alt:"Photo",height:800,width:1200,className:"next-image"}),(0,t.jsx)("h2",{children:"Differential Chromatin Accessibility and Gene Expression"}),(0,t.jsxs)("li",{children:["Results: ",(0,t.jsx)(s.a,{href:"http://trackhub.genome.duke.edu/lowelab/edotau/rabsTHREEspine/myHub/beds/differentialChromatin/",children:"Download"})]}),(0,t.jsx)("img",{src:"/pictures/chr04_30595050_30815018.svg",alt:"Photo",height:400,width:900,className:"next-image"}),(0,t.jsxs)(s.p,{children:["After genotyping 18 stickleback genomes and 4 transcriptomes, 9,468,248 SNPs genome wide were detcted. Testing each heterogygous SNPs site pinpointed to this PLXNA4 gene located near the end of chromosome 4 spanning a little more than 200 kb.\n",(0,t.jsx)("p",{children:"In the F1 hybrid WGS data is heterozygous for both SNPs. However, in the atac-seq cross, the genotype is homozygous suggesting one of the alleles is differentially increased in chromatin accessibility, while the other allele is completely closed. What is even more interesting is nearby in the rna-seq data shows gene expression is clearly down-regulated in one allele and increased gene expression with the other allele."}),"\n",(0,t.jsxs)("p",{children:["What's PLXNA4? ",(0,t.jsx)(s.a,{href:"https://www.omim.org/entry/604280",children:"PLXNA4"})," is a receptor for proteins SEMA3A and SEMA6, which mediate the effects of multiple semaphorins, including controlling diverse aspects of the nervous system development and plasticity, ranging from axon guidance and neuron migration to synaptic organization ",(0,t.jsx)(s.a,{href:"https://science.sciencemag.org/content/342/6158/1241974",children:"Sun et al., 2013"}),". PLXNA4 has recently been identified in genome wide association studies (GWAS), as a novel genetic player associated with Alzheimer's disease ",(0,t.jsx)(s.a,{href:"https://www.frontiersin.org/articles/10.3389/fnins.2018.00946/full",children:"Han et al., 2018"}),"."]}),"\n",(0,t.jsx)("h3",{children:"Raw Data"}),"\n",(0,t.jsxs)("li",{children:["ATAC-Seq: ",(0,t.jsx)(s.a,{href:"http://trackhub.genome.duke.edu/lowelab/edotau/rabsTHREEspine/raw_data/atacseq/",children:"fastqs"}),", ",(0,t.jsx)(s.a,{href:"http://trackhub.genome.duke.edu/lowelab/edotau/rabsTHREEspine/myHub/bams/atacseq/",children:"bams"}),", ",(0,t.jsx)(s.a,{href:"http://trackhub.genome.duke.edu/lowelab/edotau/rabsTHREEspine/myHub/bams/atacseq/peakCalls",children:"peaks"})]}),"\n",(0,t.jsxs)("li",{children:["RNA-Seq: ",(0,t.jsx)(s.a,{href:"http://trackhub.genome.duke.edu/lowelab/edotau/rabsTHREEspine/raw_data/atacseq/",children:"fastqs"}),", ",(0,t.jsx)(s.a,{href:"http://trackhub.genome.duke.edu/lowelab/edotau/rabsTHREEspine/myHub/bams/rnaseq",children:"bams"})," , ",(0,t.jsx)(s.a,{href:"http://trackhub.genome.duke.edu/lowelab/edotau/rabsTHREEspine/myHub/genePred",children:"genePred"}),", ",(0,t.jsx)(s.a,{href:"http://trackhub.genome.duke.edu/lowelab/edotau/rabsTHREEspine/myHub/beds/rna_seq/",children:"gene enrichments"})]}),"\n",(0,t.jsxs)("li",{children:["Genome-Seq: ",(0,t.jsx)(s.a,{href:"http://trackhub.genome.duke.edu/lowelab/edotau/rabsTHREEspine/raw_data/genomeseq",children:"fastqs"}),", ",(0,t.jsx)(s.a,{href:"http://trackhub.genome.duke.edu/lowelab/edotau/rabsTHREEspine/myHub/bams/genomeseq",children:"bams"})]})]}),(0,t.jsxs)(s.p,{children:[(0,t.jsx)("h3",{children:"Chromosome Spanning Assemblies"}),"\nHi-C libraries are commonly used today for assembling draft genomes largely because these libraries can provide insight into 3-D chomosome structures and investigate chromosome interactions in close proximity. A freshwater three-spine stickleback Hi-C library previously provided a noticeable improvement to the Gasterosteus aculeatus freshwater assembly ",(0,t.jsx)(s.a,{href:"https://academic.oup.com/jhered/article/108/6/693/3957968",children:"CL Peichel, 2017"}),"\nHowever, Karyotype characterization in sticklebacks have previously reported to have a high divergence chromosomal morphology ",(0,t.jsx)(s.a,{href:"https://www.nature.com/articles/nature08441",children:"Kitano 2009"})," , ",(0,t.jsx)(s.a,{href:"https://pubmed.ncbi.nlm.nih.gov/19229325/",children:"Ross 2009"})," When we mapped previously generated freshwater Hi-C data to both the freshwater assembly and the marine stickleback genome revealed identical heat map signatures. Similarly when we mapped marine H-C data against the freshwater and marine genomes respectfully, produced the same Hi-C heat map profiles suggesting both assemblies are high contiguity chromosome level assemblies, but no large detectable chromosomal rearrangements or translocations."]}),(0,t.jsx)("img",{src:"/pictures/hic-plot.svg",alt:"Photo",height:900,width:900,className:"next-image"}),(0,t.jsxs)(s.p,{children:[(0,t.jsx)("h3",{children:"Marine Freshwater Chomosomal Inversions"}),"\nPerforming marine to freshwater genome alignments and comparisons reveal consistent findings with previous work which identified 3 marine freshwater chromosome inversions located on chromosomes 1, 11, and 21 ",(0,t.jsx)(s.a,{href:"https://www.nature.com/articles/nature10944.pdf",children:"Jones et al. 2012"}),". On the marine stickleback genome coordinates, the inversions span roughly from 26.630.694 to 27,081,942 on chromosome one, 6,311,704 to 6,740,674 on chromosome eleven, and 9,570,312 to 11,281,238 on chromosome twenty-one."]}),(0,t.jsx)("img",{src:"/pictures/inversions.svg",alt:"Photo",height:300,width:700,className:"next-image"}),(0,t.jsxs)(s.p,{children:[(0,t.jsx)("h3",{children:"Presence Absence"}),"\nPITX1 contains multiple regulatory switches that allow for transcription of that gene in multiple tissues."]}),(0,t.jsx)("img",{src:"/pictures/pitx1_PA.svg",alt:"Photo",height:400,width:900,className:"next-image"}),(0,t.jsx)("h3",{children:"Genomic Software Projects"}),(0,t.jsxs)("li",{children:["github.com/edotau/",(0,t.jsx)(s.a,{href:"https://github.com/edotau/goFish",children:"goFish"})]}),(0,t.jsxs)("li",{children:["Slurm Compute Cluster: ",(0,t.jsx)(s.a,{href:"https://github.com/edotau/sticklebackCipher",children:"sticklebackCipher"})]}),(0,t.jsx)("h3",{children:"Lowe Lab Software"}),(0,t.jsxs)("li",{children:["Website: ",(0,t.jsx)(s.a,{href:"http://www.vertgenlab.org/",children:"vertgenlab"})]}),(0,t.jsxs)("li",{children:["github.com/ ",(0,t.jsx)(s.a,{href:"https://github.com/vertgenlab/gonomics",children:"gonomics"})]})]})}let l={MDXContent:function(){let e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:{},{wrapper:s}=Object.assign({},(0,o.a)(),e.components);return s?(0,t.jsx)(s,{...e,children:(0,t.jsx)(c,{...e})}):c(e)},pageOpts:{filePath:"pages/projects.mdx",route:"/projects",pageMap:[{kind:"MdxPage",name:"experience",route:"/experience"},{kind:"MdxPage",name:"projects",route:"/projects"},{kind:"Meta",data:{experience:"Experience",projects:"Projects"}}],flexsearch:{codeblocks:!0},title:"Projects",headings:h},pageNextRoute:"/projects",nextraLayout:i.ZP,themeConfig:r.Z};s.default=(0,n.j)(l)},14:function(e,s,a){"use strict";var t=a(5893);s.Z={logo:(0,t.jsx)("span",{children:"Home"}),project:{link:"https://github.com/edotau"},footer:{text:"edotau et. al. 2024"}}},5789:function(){}},function(e){e.O(0,[377,888,774,179],function(){return e(e.s=7468)}),_N_E=e.O()}]);